{
  "hash": "b276273d01c91d135c3383deaf284356",
  "result": {
    "markdown": "---\nauthor: \"Thadryan\"\ntitle: \"Closures in Python\"\ndate: \"2019-03-28\"\ndescription: \"Makeshift type checking in Python and when I find it handy.\"\nerror: true\n---\n\n## Closures in Python: What they are, and a practical application for type-checking objects\n\nA closure is technique that allows functions to bundle variables in their local scope for access at a later time. If there was a variable x with the value of ten in the namespace for instance, even after that code executed, we could refer back to the namespace and x would still be 10. That's a little dry; fortunately the idea is easier to illustrate in code using a function with a function inside it.\n\n::: {.cell execution_count=1}\n``` {.python .cell-code}\ndef add_ten_to_things():\n    # inside this function is a scope\n    # we'll call it \"outer scope\"\n    x = 10\n    def inner(y):\n        # this function has a scope\n        # we'll call it \"inner scope\"\n        return x + y\n    return inner\n```\n:::\n\n\nNotice how in the inner function we refer to the outer scope (to the variable x) and then return the inner scope? This is the mechanism that allows closures to work: that x is now packaged into the returned inner(). Let's see how it can be accessed:\n\n::: {.cell execution_count=2}\n``` {.python .cell-code}\n# we call the original add_ten_to_things() function that returns the inner function/namespace\nadd_ten = add_ten_to_things()\n\n# we pass something for \"y\" and it will still be able to add 10 to it\nprint(add_ten(20))\n```\n\n::: {.cell-output .cell-output-stdout}\n```\n30\n```\n:::\n:::\n\n\nBecause 10 is stored in the local namespace of the outer function that produces the new function, 10 can still be accessed by it. A function that makes functions that add ten to things isn't especially useful, however:\n\n::: {.cell execution_count=3}\n``` {.python .cell-code}\ndef add_x_to_things(x):\n    def adder(y):\n        return x + y\n    return adder\n\nincrease_by_five = add_x_to_things(5)\nincrease_by_ten = add_x_to_things(10)\n\none_plus_five = increase_by_five(1)\none_plus_ten = increase_by_ten(6)\n\nprint(one_plus_five, one_plus_ten)\n```\n\n::: {.cell-output .cell-output-stdout}\n```\n6 16\n```\n:::\n:::\n\n\nYou'd be forgiven for wondering what the point of that would be. It makes more sense once you realize you can pass arguments to the outer function and that they will be accessible to the inner one. To illustrate, consider the following example, where a function can make several types of greeter builders depending on what is passed to the outer function.\n\n::: {.cell execution_count=4}\n``` {.python .cell-code}\n# the outer function\ndef greeter_maker(greeting):\n    # inner function\n    def greeter(name):\n        print(greeting + \",\", name)\n    # outer function returns the inner \n    return greeter\n\n\nformal_greeter = greeter_maker(\"Hello, pleased to meet you\")\ncasual_greeter = greeter_maker(\"Heyo\")\n```\n:::\n\n\nIn action:\n\n::: {.cell execution_count=5}\n``` {.python .cell-code}\ncasual_greeter(\"Josh\")\nformal_greeter(\"Sruthi\")\n```\n\n::: {.cell-output .cell-output-stdout}\n```\nHeyo, Josh\nHello, pleased to meet you, Sruthi\n```\n:::\n:::\n\n\nAnother builder:\n\nWhile this show some flexibility that could be made into something useful, it can be hard to see just what a powerful idea this idea is without some context, so we will look at one more example. Consider the following class:\n\n::: {.cell execution_count=6}\n``` {.python .cell-code}\nclass Person(object):\n    def __init__(self, name, age):\n        self.name = name\n        self.age = age\n```\n:::\n\n\nThis is an extremely simple Python class that models a Person with a name and age. It's probably obvious to any hypothetical user of this class that \"name\" is a string and \"age' is an integer (or at least a number or some kind). However, as Python is dynamically typed, there is nothing in the language itself to enforce this. Consider the following:\n\n::: {.cell execution_count=7}\n``` {.python .cell-code}\n# everything goes according to plan...\nchristian = Person(\"Christian Slater\", 49)\n\n# whoops, wrong order...\nrami = Person(37, \"Rami Malek\")\n\n# gets nane \nprint(christian.name)\n# not what we though we'd get\nprint(rami.name)\n```\n\n::: {.cell-output .cell-output-stdout}\n```\nChristian Slater\n37\n```\n:::\n:::\n\n\nThis program will run just fine until sometime downstream when someone tries to do a computation a involving rami's name or age but has the wrong type of data (ie, age += 1 and finds it's a string). Consider the following usage of a closure to prevent this (this is much simpler than it seems once you see it in context so hang on):\n\n::: {.cell execution_count=8}\n``` {.python .cell-code}\n# takes a dictionary of attributes->types, and a model\ndef typechecking_builder(required_params, model):\n    \n    def _builder(attributes_passed):\n        # iterate through the dict that was passed to the outer function\n        for k in attributes_passed.keys():\n            current_param = attributes_passed[k]\n            type_required = required_params[k]\n            # ensure they're what we've been told to expect\n            try:\n                assert type(current_param) == type_required\n                print(\"Type check passed for\", current_param, \"of type\", type_required)\n            # raise an error if they're not\n            except AssertionError:\n                print(\"\\tfound\", current_param, \"of type\", type(current_param), \"Required:\", type_required)\n                raise TypeError(\"Type check failed for:\", k) # whatever you want to do for exception handling\n    # return the type checking function\n    return _builder\n```\n:::\n\n\nWe can now make a builder that will type check for us, telling the function what it should expect when it is called:\n\n::: {.cell execution_count=9}\n``` {.python .cell-code}\n# make people, thier name is a string, thier age is an int, they're of type Person\nperson_builder = typechecking_builder({\"name\": str, \"age\":int}, Person)\n```\n:::\n\n\nWhen we use this, it will check the types we passed:\n\n::: {.cell execution_count=10}\n``` {.python .cell-code}\ncarly = person_builder({\"name\": \"Carly Chaikin\", \"age\": 28})\n```\n\n::: {.cell-output .cell-output-stdout}\n```\nType check passed for Carly Chaikin of type <class 'str'>\nType check passed for 28 of type <class 'int'>\n```\n:::\n:::\n\n\nNow, if we make the simple mistake from before, we get feedback right away:\n\n::: {.cell execution_count=11}\n``` {.python .cell-code}\nsunita = person_builder({\"name\": 32, \"age\": \"Sunita Mani\"})\n```\n\n::: {.cell-output .cell-output-stdout}\n```\n\tfound 32 of type <class 'int'> Required: <class 'str'>\n```\n:::\n\n::: {.cell-output .cell-output-error}\n```\nTypeError: ('Type check failed for:', 'name')\n```\n:::\n:::\n\n\n...and every Person we build will be type checked!\n\n",
    "supporting": [
      "index_files"
    ],
    "filters": [],
    "includes": {}
  }
}